### PR TITLE
feat: add new endpoint to get the progress of a dataset

### LINKED ISSUE #91: [TASK] Add endpoint to return a dataset progress

### PR DESCRIPTION
**Pull request by jfcalvo**
# Description

This PR add a new endpoint with the following signature:

`GET /api/v1/datasets/:dataset_id/progress`

This endpoint will return a response with the following format:

```json
{
  "total": 8,
  "submitted": 2,
  "discarded": 2,
  "conflicting": 1,
  "pending": 3,
}
```

The attributes will return an integer with the following logic:
* `total`: the total number of records for the specified dataset.
* `submitted`: the number of records for the specified dataset with at least one response with `submitted` status and no responses with `discarded` status.
* `discarded`: the number of records for the specified dataset with at least one response with `discarded` status and no responses with `submitted` status.
* `conflicting`: the number of records for the specified dataset with at least one response with `submitted` status and at least one response with `discarded` status.
* `pending`: the number of pending records. This is calculated like `pending = total - submitted - discarded - conflicting`.

Closes #91 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (change restructuring the codebase without changing functionality)
- [ ] Improvement (change adding some improvement to an existing functionality)
- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [x] Adding new tests.

**Checklist**

- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla_server/contexts/datasets.py [hunk 1]
```diff
@@ -449,6 +450,40 @@ async def count_records_by_dataset_id(db: "AsyncSession", dataset_id: UUID) -> i
...
+    total = await count_records_by_dataset_id(db, dataset_id)
+    submitted = (
+        await db.execute(select(func.count("*")).select_from(submitted_query.except_(discarded_query)))
+    ).scalar_one()
+    discarded = (
+        await db.execute(select(func.count("*")).select_from(discarded_query.except_(submitted_query)))
+    ).scalar_one()
+    conflicting = (
+        await db.execute(select(func.count("*")).select_from(submitted_query.intersect(discarded_query)))
+    ).scalar_one()
```
[the remark this answers is not in the document]
Comment (frascuchon) [reply src/argilla_server/contexts/datasets.py:475]: please proceed :-)

**Review by frascuchon — APPROVED** [review]
hoping to find some endpoint improvement

File: src/argilla_server/contexts/datasets.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jfcalvo) [reply src/argilla_server/contexts/datasets.py]: Done!

File: src/argilla_server/contexts/datasets.py [hunk 2]
```diff
@@ -449,6 +447,77 @@ async def count_records_by_dataset_id(db: "AsyncSession", dataset_id: UUID) -> i
     return (await db.execute(select(func.count(Record.id)).filter_by(dataset_id=dataset_id))).scalar_one()
 
 
+async def get_dataset_progress_with_group_by(db: "AsyncSession", dataset_id: UUID) -> DatasetProgress:
```
Comment (jfcalvo) [inline src/argilla_server/contexts/datasets.py:450]: We will use @frascuchon implementation because is faster.

**Review by frascuchon — APPROVED** [review]

**Pull request merged**